outbox-kafka-spring/src/main/java/one/tomorrow/transactionaloutbox/service/OutboxProcessor.java

```diff
@@ -165,15 +163,31 @@ private void processOutboxWithLock() {
 
 	}
 
-	private void processOutbox() throws ExecutionException, InterruptedException {
-		List<OutboxRecord> records = repository.getUnprocessedRecords(BATCH_SIZE);
-		for (OutboxRecord outboxRecord : records) {
-			ProducerRecord<String, byte[]> producerRecord = toProducerRecord(outboxRecord);
-			Future<RecordMetadata> result = producer.send(producerRecord);
-			result.get();
-			logger.info("Sent record to kafka: {}", outboxRecord);
-			outboxRecord.setProcessed(now());
-			repository.update(outboxRecord);
+	private void processOutbox() {
+		repository.getUnprocessedRecords(BATCH_SIZE)
+				.stream()
+				.map(outboxRecord -> {
+					ProducerRecord<String, byte[]> producerRecord = toProducerRecord(outboxRecord);
+					return Tuple.of(outboxRecord, producer.send(producerRecord));
+				})
+				.toList() // collect to List before awaiting futures, to use producer internal batching
+				.forEach(tuple -> {
+					await(tuple.b);
+					OutboxRecord outboxRecord = tuple.a;
+					logger.info("Sent record to kafka: {}", outboxRecord);
+					outboxRecord.setProcessed(now());
+					repository.update(outboxRecord);
+				});
```

Comment:
A applied your proposal slightly differently:
* take exception into account
* don't serialize processing of items, but map over all items loaded from the db before awaiting the first record 

Because `xs.map(f1).forEach(f2)` invokes `f2` for the result of `f1(xs[0])` before invoking `f1(xs[1])` - maybe that's not obvious. I.e. your proposal would still have serialized the processing and would have been only a different syntax for the same semantics as the original code.

WDYT?

Reply:
Btw, I just added a unit test verifying the expected behavior - commit 13140f8f421144ad0b3d397924a4f9149d6c31d8

Reply:
In my proposal the exception was logged on error level, i see warning as a good solution as well, as they will be retried until they have been processed, so hopefully should heal themselves.

Good catch that I missed the to list in my proposal, of course we need to keep it to keep processing parallel.

LGTM